Add SignalR function extension

Y-Sindo opened the pull request:
- The first commit of SignalR extension
- Add ci.yml

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers that changes made by the pull request.

If you are regenerating your SDK based off of a new swagger spec, please add the link to the corresponding swagger spec pull request that has been merged in the azure-rest-api-specs repository
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

Y-Sindo committed:
The first commit of SignalR extension

Y-Sindo committed:
Add ci.yml

Y-Sindo commented:
/azp run prepare-pipelines

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/.gitignore:
```diff
@@ -0,0 +1,362 @@
+## Ignore Visual Studio temporary files, build results, and
```
Most of these seem pretty generic and should already be covered in the repo's main gitignore - https://github.com/Azure/azure-sdk-for-net/blob/main/.gitignore

Can you check to see if this would cover everything here, or if there are SignalR specific entries here?

Y-Sindo replied:
The only thing not covered by main gitignore is "local.settings.json". Can "local.settings.json" be added to main gitignore? I think that might be helpful for all the function extension projects with function samples.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/Directory.Build.props:
```diff
@@ -0,0 +1,25 @@
+﻿<Project>
```
I think we should standardize this to match other libraries, e.g. https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Directory.Build.props

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/Directory.Build.targets:
```diff
@@ -0,0 +1,6 @@
+﻿<Project>
+  <PropertyGroup>
```
This file can be deleted.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/README.md:
```diff
@@ -0,0 +1,123 @@
+# Azure Functions Bindings for Azure SignalR Service
```
Can you take a look at the format used in https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/README.md and match it?

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/binding-library/java/README.md:
```diff
@@ -0,0 +1,51 @@
+# SignalR Service Library for Azure Java Functions
```
I don't think we would want to host Java code here. I'm not sure we would even be able to publish it from this repo.
@AlexGhiondea @weshaggard Should we consider moving this to the Java repo?

weshaggard replied:
I agree we shouldn't put java code here. Do you guys have a java SDK, or is this that? In either case as Josh calls out we cannot build or publish any java code from this repo.

Y-Sindo replied:
The java code hasn't been updated for a long time. I think I can just remove it. If needed later, I will move it to Java repo.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/samples/bidirectional-chat/csharp/.gitignore:
```diff
@@ -0,0 +1,264 @@
+## Ignore Visual Studio temporary files, build results, and
```
This can probably be removed in favor of the main gitignore

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/samples/chat-with-auth/csharp/FunctionApp/.gitignore:
```diff
@@ -0,0 +1,264 @@
+## Ignore Visual Studio temporary files, build results, and
+## files generated by popular Visual Studio add-ons.
+
+# Azure Functions localsettings file
```
Can probably remove

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/samples/chat-with-custom-auth/csharp/FunctionApp/.gitignore:
```diff
@@ -0,0 +1,264 @@
+## files generated by popular Visual Studio add-ons.
+
+# Azure Functions localsettings file
+local.settings.json
```
Can probably remove this file as well.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/version.props:
```diff
@@ -0,0 +1,11 @@
+<Project>
+  <PropertyGroup>
+    <VersionPrefix>1.6.0</VersionPrefix>
```
I don't think this would be needed anymore.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/test/Microsoft.Azure.Webjobs.Extensions.SignalRService.E2ETests/Microsoft.Azure.Webjobs.Extensions.SignalRService.E2ETests.csproj:
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Take a look at https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/template/Azure.Template/tests to see how we structure/name our test projects/directories.

If it isn't too much of a change, it might be worth following the same pattern.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalRServiceExtension/TriggerBindings/Utils/SignalRTriggerUtils.cs:
```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// Copyright (c) Microsoft. All rights reserved.
```
nit: repeated file header.

JoshLove-msft commented:
/azp run prepare-pipelines

JoshLove-msft commented:
/azp run net - signalr - ci

Y-Sindo committed:
fix

Y-Sindo committed:
fix .gitignore

Y-Sindo committed:
Remove Java code

Y-Sindo committed:
Standardize `Directory.Build.props` and dependencies

Y-Sindo committed:
Remove E2E tests

Will add them later.

Y-Sindo committed:
Merge branch 'main' into signalr

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/README.md:
```diff
@@ -0,0 +1,123 @@
+
+## Build Status
+
+Travis: [![travis](https://travis-ci.org/Azure/azure-functions-signalrservice-extension.svg?branch=dev)](https://travis-ci.org/Azure/azure-functions-signalrservice-extension)
```
Should this be removed?

Y-Sindo replied:
I will try to rewrite the readme in the format of ServiceBus.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/azure-functions-signalrservice-extension.sln:
```diff
@@ -0,0 +1,130 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
```
I'd suggest renaming the solution to match the convention of other WebJobs solutions.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/samples/simple-chat/java/src/main/java/com/functions/Functions.java:
```diff
@@ -0,0 +1,72 @@
+import com.microsoft.azure.functions.signalr.annotation.*;
+import com.microsoft.azure.functions.signalr.*;
+
+public class Functions {
```
@weshaggard is it alright to have Java samples in the repo? They aren't expected to build (at least at this point we are not building samples).

weshaggard replied:
I would suggest we do not have java samples in the .NET repo. I assume these samples would need to to be calling java libraries so they should live in the java repo with the java library.

JoshLove-msft replied:
> I assume these samples would need to to be calling java libraries so they should live in the java repo with the java library.

There aren't really Java equivalent libraries to call - the .NET code is used via an out of process mechanism.

Y-Sindo replied:
I think we could move the Java, Js sample codes to https://github.com/aspnet/AzureSignalR-samples/

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/samples/simple-chat/js/functionapp/addToGroup/index.js:
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+module.exports = function (context, req) {
```
@weshaggard similar question around having JS samples.

weshaggard replied:
Similar answer they should live next to the JS library in the JS repo.

Y-Sindo committed:
Remove js,java sample codes and rename directories

Y-Sindo committed:
Rename .sln file

Y-Sindo committed:
Rename `src/SignalRServiceExtension` to `src/Microsoft.Azure.WebJobs.Extensions.SignalRService`

Y-Sindo committed:
Rename `test/SignalRServiceExtension.Tests` to `test/Microsoft.Azure.WebJobs.Extensions.SignalRService.Tests`

Y-Sindo committed:
Rename `test` to `tests`

Y-Sindo committed:
Remove unused items in sln file

Y-Sindo committed:
Merge branch 'signalr' of https://github.com/Y-Sindo/azure-sdk-for-net into signalr

Y-Sindo committed:
Fix Protocols.csproj

Y-Sindo committed:
Many fix

Y-Sindo committed:
Remove Serverless protocol projects and move other codes

Y-Sindo committed:
Fix target frameworks

weshaggard commented on eng/Packages.Data.props:
```diff
@@ -108,9 +108,11 @@
-    <PackageReference Update="Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson" Version="3.0.0" Condition="'$(TargetFramework)' == 'netcoreapp3.1'"/>
-    <PackageReference Update="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
-    <PackageReference Update="Microsoft.Azure.SignalR.Management" Version="1.11.0" />
+    <PackageReference Update="MessagePack" Version="1.9.11" />
```
What is the package? We need to approve this before we can depend on it in one of our shipping libraries.

Y-Sindo replied:
According to subsequent modifications, MessagePack is not needed.
`Microsoft.AspNetCore.SignalR.Protocols.MessagePack` is used for serialization. It is used for tests now and I could move it to dependencies for test projects.
`Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson` is also for serialization. It is needed by SignalR extensions.
`Microsoft.Azure.SignalR.Management` is a core dependency used by SiganlR extension, hosted in https://github.com/Azure/azure-signalr/.
`Microsoft.Azure.SignalR.Serverless.Protocols` is hosted in the [old SiganlR extension repo](https://github.com/Azure/azure-functions-signalrservice-extension/), and might need to move to this repo later.

weshaggard replied:
If possible lets remove the dependency even from tests. We want to avoid any none Microsoft/Azure based packages dependences in the repo as much as possible.

Y-Sindo replied:
Currently they are all Microsoft/Azure based packages. But if we also move `Microsoft.Azure.SignalR.Serverless.Protocols` this project here, then seems `MessagePack` is unavoidable.

weshaggard replied:
Regardless of where Microsoft.Azure.SignalR.Serverless.Protocols lives we need to either get the that package dependency approved by the ArchBoard or we need to remove it.

Y-Sindo committed:
Try fix code in srcs

The problem is the $(RequiredTargetFrameworks) is netstandard2.0.   But I need .net core 3.1.

Y-Sindo committed:
Add SignalR nuget source for dev

Y-Sindo committed:
Fix test project

Y-Sindo committed:
Fix nuget source problem

Y-Sindo committed:
Update `Packages.Data.props`

Y-Sindo committed:
Remove `Management` project from `ci.yml`

Y-Sindo committed:
Add `service.projects`

Y-Sindo committed:
Remove samples

Y-Sindo committed:
Add version and CHANGELOG.md

Y-Sindo committed:
Fix docs

Y-Sindo committed:
Fix readme.
Add comparison of concept **SignalR Service client** and **SignalR client**

Y-Sindo committed:
Export public APIs

Y-Sindo committed:
Fix code snippets in readme

Y-Sindo committed:
Fix warnings for src/*

Y-Sindo committed:
Fix warnings for test/*

Y-Sindo committed:
Add XML doc for public types

Y-Sindo committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into signalr

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/18274553e71a4f44a2f924031f627b14?diffOnly=True&diffRevisionId=9e7de7b53f1e43bda1bad21456d16f97)

Y-Sindo commented:
> API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/18274553e71a4f44a2f924031f627b14?diffOnly=True&diffRevisionId=9e7de7b53f1e43bda1bad21456d16f97)

The class `SignalROptions` is removed in 1.3.0, but the base API version is 1.2.2. I would like to know how the base version is computed.

Y-Sindo committed:
Remove unused file

Y-Sindo committed:
Add code snippet for message trigger

Y-Sindo committed:
remove gitignore file

Y-Sindo committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into signalr

Y-Sindo committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into signalr
